examples/directchat.nim

```diff
@@ -0,0 +1,224 @@
+
+# forward declaration
+proc readWriteLoop(p: ChatProto) {.async, gcsafe.}
+proc readAndPrint(p: ChatProto) {.async, gcsafe.} =
+  while true:
+    while p.connected:
+      # echo &"{p.id} -> "
+      echo cast[string](await p.conn.readLp())
+    await sleepAsync(100.millis)
+
+proc dialPeer(p: ChatProto, address: string) {.async, gcsafe.} =
+  var parts = address.split("/")
+  if parts.len == 11 and parts[^2] != "ipfs" and parts[^2] != "p2p":
+    quit("invalid or incompelete peerId")
+
+  var remotePeer: PeerInfo
+  remotePeer.peerId = some(PeerID.init(parts[^1]))
+  remotePeer.addrs.add(MultiAddress.init(address))
+
+  echo &"dialing peer: {address}"
+  p.conn = await p.switch.dial(remotePeer, ChatCodec)
+  p.connected = true
+
+proc writeAndPrint(p: ChatProto) {.async, gcsafe.} = 
+  while true:
+    if not p.connected:
+      # echo &"{p.id} ->"
+    # else:
+      echo "type an address or wait for a connection:"
+      echo "type /[help|?] for help"
+
+    var line = await p.transp.readLine()
+    if line.startsWith("/help") or line.startsWith("/?") or not p.started:
+      echo Help
+      continue
+  
+    if line.startsWith("/disconnect"):
+      echo "Ending current session"
+      if p.connected and p.conn.closed.not:
+        await p.conn.close()
+      p.connected = false
+    elif line.startsWith("/connect"):
+      if p.connected:
+        var yesno = "N"
+        echo "a session is already in progress, do you want end it [y/N]?"
+        yesno = await p.transp.readLine()
+        if yesno.cmpIgnoreCase("y") == 0:
+          await p.conn.close()
+          p.connected = false
+        elif yesno.cmpIgnoreCase("n") == 0: 
+          continue
+        else:
+          echo "unrecognized response"
+          continue
+
+      echo "enter address of remote peer"
+      let address = await p.transp.readLine()
+      if address.len > 0:
+        await p.dialPeer(address)
+
+    elif line.startsWith("/exit"):
+      if p.connected and p.conn.closed.not:
+        await p.conn.close()
+        p.connected = false
+
+      await p.switch.stop()
+      echo "quitting..."
+      quit(0)
+    else:
+      if p.connected:
+        await p.conn.writeLp(line)
+      else:
+        try:
+          if line.startsWith("/") and "ipfs" in line:
+            await p.dialPeer(line)
+        except:
+          echo &"unable to dial {line}"
+          # echo getCurrentExceptionMsg()
+
+proc readWriteLoop(p: ChatProto) {.async, gcsafe.} =
+    asyncCheck p.writeAndPrint()
+    asyncCheck p.readAndPrint()
+
+method init(p: ChatProto) {.gcsafe.} =
+  proc handle(stream: Connection, proto: string) {.async, gcsafe.} = 
+    if p.connected and not p.conn.closed:
+      echo "a chat session is already in progress - disconnecting!"
+      await stream.close()
+
+    p.conn = stream
+    p.connected = true
+
+  p.codec = ChatCodec
+  p.handler = handle
+
+proc newChatProto(switch: Switch, transp: StreamTransport): ChatProto = 
+  new result
```

Comment:
same here - can use object syntax:
```
result = ChatProto(switch: ..., transp: ...)
```

generally, we also pass arguments to `init` and have it construct the object fully, this is in anticipation of a generic init strategy that's being added to the language (so `init` returns the instance like a constructor)

Reply:
Would it be something similar to this pattern - https://github.com/status-im/nim-beacon-chain/blob/9181df9277/beacon_chain/block_pool.nim#L40-L131?

Reply:
yeah.

ideally, we try to avoid inits by constructing the types such that the default value (0/nil) is meaningful enough that no special init is needed, but for those cases where it is...

this is why `option[t]` has default "not-set", so that when stuff is default-constructed it is still meaningful and the "safer" choice

Reply:
will be addressed in upcoming (more thorough) cleanup